src/v/cluster/persisted_stm.cc

```diff
@@ -224,8 +224,18 @@ ss::future<> persisted_stm::start() {
 
     if (maybe_snapshot) {
         stm_snapshot& snapshot = *maybe_snapshot;
-        co_await apply_snapshot(snapshot.header, std::move(snapshot.data));
-        set_next(raft::details::next_offset(_last_snapshot_offset));
+
+        if (snapshot.header.offset >= _c->start_offset()) {
```

Comment:
I have one question here, is it safe to skip reading the snapshot ? In other words is it enough for transaction related state machines to recover using information available in log ? (some of the records may be lost because of the retention policy). I am wondering if those state machines require those information, which should be present in a snapshot ?

Reply:
I planned for this scenario. We store snapshot instead of the updates in the tx / idempotency logs (rm_stm & tm_stm) and use LWW conflict resolution strategy so where ever we make a cut (prefix deletion) we'll observe only committed data. See this commit message for more information - https://github.com/vectorizedio/redpanda/commit/05c94d41209bc189a75597eec1bddfe1f8dc168a.

With uncommitted situation out of question the only concern is that we don't see all committed data and lose data. It's ok. When during an idempotency check Redpanda can't match current sequence number with the stored sequence number it throws OutOfOrderSequence error so user re-initialize a session and this event doesn't depend on the historic data.

Similar thing happens with the transactions. When a client tries to act on the lost transaction they gets a reasonable error and roll back the transaction. But the general idea is that a transaction times out before deletion.